### PR TITLE
Some Transformers Do Not Have Telemetry Trigger

### PR DESCRIPTION
**Pull request by BackendBananeWala**
Fixes: https://github.com/BharatSahAIyak/packages/issues/91
Fixes: https://github.com/BharatSahAIyak/packages/issues/149

## Summary by CodeRabbit

- **New Features**
	- Introduced a `TelemetryLogger` to various transformer classes for enhanced telemetry logging capabilities.
	- Added telemetry logging at the start and end of transformation processes in several transformers, improving monitoring and error tracking.

- **Bug Fixes**
	- Improved error message logging during transformation failures, providing better insights into issues encountered.

- **Tests**
	- Enhanced test suites for transformers by introducing a centralized `mockConfig` for consistency and clarity in test setups.
	- Added verification for telemetry logging calls in tests to ensure correct event tracking during transformations.

### CODE REVIEW

File: packages/transformers/src/modules/generic/field_setter/field_setter.transformer.spec.ts [hunk 1]
```diff
@@ -160,14 +166,18 @@ describe('Field Setter Tests', () => {
...
         const transformedMsg = await transformer.transform(mockXMessage);
         xmsgCopy.payload.text = '';
         transformedMsg.payload.text = '';
         transformedMsg.channelURI = '';
         xmsgCopy.channelURI = '';
-        expect(xmsgCopy).toStrictEqual(transformedMsg);
+        console.log(transformedMsg);
+        console.log(xmsgCopy);
+        // There is an additional field in the transformed message which is not present in the original message
+        expect(transformedMsg).toMatchObject(xmsgCopy);
```
Comment (chinmoy12c) [inline packages/transformers/src/modules/generic/field_setter/field_setter.transformer.spec.ts:180]: What field is that exactly? Ideally, I would not expect any field to change apart from the ones listed just above this.

File: packages/transformers/src/modules/generic/speect_to_text/speech_to_text.transformer.spec.ts [hunk 2]
```diff
@@ -7,8 +7,16 @@ import {
...
 import { SpeechToTextTransformer } from "./speech_to_text.transformer";
 import FormData from "form-data";
+import { TelemetryLogger } from "../../common/telemetry";
 
 describe("SpeechToTextTransformer", () => {
+  let mockEvent = { pushEvent: jest.fn() };
+  let mockConfig = { eventBus: mockEvent, transformerId: "speech-to-text-transformer" };
+  // let mockLogger = { sendErrorTelemetry: jest.fn(), sendLogTelemetry: jest.fn() };
+  // jest.spyOn(TelemetryLogger.prototype, "sendErrorTelemetry").mockImplementation(mockLogger.sendErrorTelemetry);
+  // jest.spyOn(TelemetryLogger.prototype, "sendLogTelemetry").mockImplementation(mockLogger.sendLogTelemetry);
```
Comment (chinmoy12c) [inline packages/transformers/src/modules/generic/speect_to_text/speech_to_text.transformer.spec.ts:17]: We should remove commented out code if it not being used.

File: packages/transformers/src/modules/generic/speect_to_text/speech_to_text.transformer.spec.ts [hunk 3]
```diff
@@ -92,6 +105,10 @@ describe("SpeechToTextTransformer", () => {
         bot: true,
         meta: new Map(Object.entries({ botMobileNumber: "919999999999" })),
       },
+      transformer: {
+        metaData: {
+        },
+      },
```
Comment (chinmoy12c) [inline packages/transformers/src/modules/generic/speect_to_text/speech_to_text.transformer.spec.ts:111]: What was the purpose of adding this?

File: packages/transformers/src/modules/if_else/random_binary/random_binary.transformer.ts [hunk 4]
```diff
@@ -1,9 +1,11 @@
 import { XMessage } from "@samagra-x/xmessage";
 import { ITransformer } from "../../common";
+import { TelemetryLogger } from "../../common/telemetry";
 
 export class RandomBinaryTransformer implements ITransformer {
 
     constructor(readonly config: Record<string, any>) { }
+    private readonly telemertyLogger = new TelemetryLogger(this.config);
```
Comment (chinmoy12c) [inline packages/transformers/src/modules/if_else/random_binary/random_binary.transformer.ts:8]:
```suggestion
    private readonly telemetryLogger = new TelemetryLogger(this.config);
```

File: packages/transformers/src/modules/retry/simple_retry/simple_retry.transformer.ts [hunk 5]
```diff
@@ -1,15 +1,20 @@
...
     ///     retries: string: Number of reties before failure. Default is 1. (optional)
     ///     delay: number: Delay in milliseconds before next retry. Default is 0. (optional)
     constructor(readonly config: Record<string, any>) { }
+    private readonly telemetryLogger = new TelemetryLogger(this.config);
+
 
     async transform(xmsg: XMessage): Promise<XMessage> {
+        this.telemetryLogger.sendLogTelemetry(xmsg, `${this.config.transformerId} started!`, Date.now());
         if (!xmsg.transformer) {
+            this.telemetryLogger.sendLogTelemetry(xmsg, 'Transformer metadata not found. Creating new metadata.', Date.now());
```
Comment (chinmoy12c) [inline packages/transformers/src/modules/retry/simple_retry/simple_retry.transformer.ts:17]: we can leave this out

File: packages/transformers/src/modules/switch_case/label_classifier/label_classifier.transformer.ts [hunk 6]
```diff
@@ -19,13 +18,13 @@ export class LabelClassifierTransformer implements ITransformer{
     ///     minimumThreshold: number: If provided, a label must reach this threshold score to be considered for final result. Default is 0. (optional)
     ///
     ///     Note: `existingLabel` can also be passed in XMessage `metaData.existingLabel`.
-    constructor(config: Record<string, any>) {
-        this.config = config;
-    }
+    constructor(public readonly config: Record<string, any>){ }
```
Comment (chinmoy12c) [inline packages/transformers/src/modules/switch_case/label_classifier/label_classifier.transformer.ts:21]: Please revert this change. AKAIK this creates issues with interface compatibility.

**Review by chinmoy12c — CHANGES_REQUESTED** [review]
You also need to update version in `package.json`.

File: packages/transformers/src/modules/generic/field_setter/field_setter.transformer.spec.ts (hunk 1 above)
Comment (BackendBananeWala) [reply packages/transformers/src/modules/generic/field_setter/field_setter.transformer.spec.ts:180]: When we add Telemetry on the transformers they change the object and in meta they add three keys namely 
`telemetryLog
stateExecutionTime
errorString`

File: packages/transformers/src/modules/generic/speect_to_text/speech_to_text.transformer.spec.ts (hunk 2 above)
Comment (BackendBananeWala) [reply packages/transformers/src/modules/generic/speect_to_text/speech_to_text.transformer.spec.ts:17]: Done

File: packages/transformers/src/modules/generic/speect_to_text/speech_to_text.transformer.spec.ts (hunk 3 above)
Comment (BackendBananeWala) [reply packages/transformers/src/modules/generic/speect_to_text/speech_to_text.transformer.spec.ts:111]: As they are missing the **telemetry** was not able to add the meta data as listed above

File: packages/transformers/src/modules/retry/simple_retry/simple_retry.transformer.ts (hunk 5 above)
Comment (BackendBananeWala) [reply packages/transformers/src/modules/retry/simple_retry/simple_retry.transformer.ts:17]: Done

File: packages/transformers/src/modules/switch_case/label_classifier/label_classifier.transformer.ts (hunk 6 above)
Comment (BackendBananeWala) [reply packages/transformers/src/modules/switch_case/label_classifier/label_classifier.transformer.ts:21]: Fixed

File: packages/transformers/src/modules/generic/llm/llm.transformer.ts [hunk 7]
```diff
@@ -174,6 +176,10 @@ export class LLMTransformer implements ITransformer {
         }
 
         xmsg.messageId.replyId = xmsg.messageId.Id;
+        this.switchFromTo(xmsg);
+        const oldMessageId = xmsg.messageId.Id;
+        const newMessageId = uuid4();
+        xmsg.messageId.Id = newMessageId;
```
Comment (chinmoy12c) [inline packages/transformers/src/modules/generic/llm/llm.transformer.ts:182]: please revert this, it changes the behaviour of the transformer

**Pull request merged**